enhancements/kube-apiserver/stale-clusteroperators-detection.md

```diff
@@ -0,0 +1,108 @@
+
+## Proposal
+
+To implement stale operator detection, we need a controller process that will run close to the Kubernetes API server,
+preferably using a local connection and client-based certificate for authentication to avoid dependencies on working
+auth and network when the cluster is unstable.
+
+The controller will watch the clusteroperator resource and check the `status.[]conditions.lastTransitionTime` field for
+Degraded, Available, Progressing, and Upgradeable conditions. Suppose the "lastTransitionTime" field is not updated for
+10 minutes. In that case, the controller will change the human-readable condition message
+to: `Operator checking for stale status, the active operator will reset this message: <original message>`.
+
+After the message is changed, as it suggests, the active operators should reset this message back to the original
+message, removing the "checking" part. This shows a healthy, active operator.
+
+Suppose the operator is not active and does not reset the message back to the original after 10 minutes. In that case,
+the controller will change the message again
+to: `Operator has not updated this condition for more than 20 minutes, last known condition state was "<False/True>", original message: <original message>`
+. In addition to changing the message, this time, the controller will also change the condition status from False or
+True to **Unknown**.
+
+Since the clusteroperator resource has multiple conditions in status, if **any** of these conditions are updated, we assume
+the operator is still active and able to reflect the operand status into clusteroperator resource.
+
+For cluster administrators, this means the operator is having a problem and is potentially down or unable to make
+updates to conditions. In that case, an alert might be required to trigger a manual investigation of the current
+operator state.
+
+### API Extensions
+
+*No API extensions required*
+
+### Risks and Mitigations
+
+* One risk here is that we find operators that are not correctly reconciling their clusteroperator conditions, which this controller makes obvious. These operators should be fixed as a result.
+* This also creates (2 writes) * (number of cluster operators) every N minutes
+
+## Design Details
+
+### Test Plan
+
+* This feature should be pared with an e2e invariant that ensures a minimum update time for the stale message. Let's take a guess at 30s to start and see how many fail.
+
+### Operational Aspects of API Extensions
+
+No API extension is needed to implement this proposal.
+
+#### Failure Modes
+
+#### Support Procedures
+
+## Implementation History
+
+## Drawbacks
+
+## Alternatives
+
+One alternative discussed was a new API field that will be added to clusteroperators.Status called "lastUpdateTime".
+Instead of relying on
+"lastTransitionTime" will introduce this field and use it as "heart-beat". This will require **all** operators to update
+this field correctly and multiply the writes significantly.
```

Comment:
not clear to me why this would increase the number of writes?  In fact i think it would potentially reduce the number of writes.

the current proposal requires 2x writes every 10 minutes, per operator (one to set the message as potentially stale, another to set it back to the correct value)

a new field would only require that each operator poke the field every 10 minutes, so 1x write every 10 minutes (the heartbeat controller would have to watch all the operators, so potentially 1x read every 10 minutes)

but yes, it would require all operators to add new logic to maintain the new field.

Note that this alternate approach would also have the benefit of eliminating the "Flapping status message" behavior that i believe is built into the proposed approach.

Reply:
Conditions are standardized across in all of kube in apimachinery metav1.  Consumers of messages expect the schemas to match.  Our previous heartbeats were set at the nyquist frequency, so it was always 2x the speed of your desired detection limit..

Keeping the schemas consistent for external handling appears more important than having a stable message.

Reply:
> Keeping the schemas consistent for external handling appears more important than having a stable message.

This seems to be a subjective value judgement.  Appears more important based on what, exactly?  Even taken on its face, having a core consistent subset of conditions that all external handlers can expect and must understand seems valuable, but claiming no extensions can be made that external handlers are free to ignore seems a little more difficult to defend.

But also, this alternative proposal doesn't suggest a new condition anyway, it suggests a new status field entirely.

Reply:
How averse are we to requiring updates to existing cluster operators as part of this proposal?

Without changing the API, or even introducing a new condition type, I could imagine the staleness controller comparing the ObservedGeneration of conditions and probing via something that causes an increment to metadata.generation, say, an inert annotation (set to the metadata.generation observed by the staleness controller at the time of the probe?).